File: orchestration/src/tasks/helpers/rosbag_profiles.py
from __future__ import annotations

import re
from pathlib import Path
from typing import Any


def profile_topics(profile_path: Path) -> tuple[list[str], list[str], list[str]]:
    required: list[str] = []
    optional: list[str] = []
    if profile_path.is_file():
        for raw_line in profile_path.read_text(encoding="utf-8").splitlines():
            line = raw_line.strip()
            if not line or line.startswith("#"):
                continue
            parts = line.split(maxsplit=1)
            if len(parts) != 2:
                continue
            kind, topic = parts
            if kind == "required":
                required.append(topic)
            elif kind == "optional":
                optional.append(topic)
    return required, optional, [*required, *optional]


def load_rosbag_metadata_counts(metadata: Path) -> dict[str, int]:
    content = metadata.read_text(encoding="utf-8")
    counts: dict[str, int] = {}
    topic_matches = list(re.finditer(r"name: (/[^\n]+)", content))
    for index, match in enumerate(topic_matches):
        topic = match.group(1).strip()
        end = topic_matches[index + 1].start() if index + 1 < len(topic_matches) else len(content)
        block = content[match.end() : end]
        count_match = re.search(r"message_count:\s*(\d+)", block)
        counts[topic] = int(count_match.group(1)) if count_match else 0
    return counts


def validate_official_rosbag_profile(
    *,
    profile: Path,
    metadata: Path,
    required: list[str],
    optional: list[str],
) -> dict[str, Any]:
    counts = load_rosbag_metadata_counts(metadata)
    missing_required = [topic for topic in required if topic not in counts]
    zero_count_required = [topic for topic in required if counts.get(topic, 0) <= 0]
    present_optional = [topic for topic in optional if counts.get(topic, 0) > 0]
    missing_optional = [topic for topic in optional if topic not in counts]
    return {
        "ok": not missing_required and not zero_count_required,
        "recorded": True,
        "profile": str(profile),
        "metadata": str(metadata),
        "required_topics": required,
        "optional_topics": optional,
        "present_topics": sorted(counts),
        "message_counts": counts,
        "missing_required_topics": missing_required,
        "zero_count_required_topics": zero_count_required,
        "present_optional_topics": present_optional,
        "missing_optional_topics": missing_optional,
    }

File: orchestration/sim/internal/tasks/helpers/rosbag_profiles.go
package helpers

import (
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

type Topics struct {
	Required []string
	Optional []string
	All      []string
}

type Validation struct {
	OK                      bool           `json:"ok"`
	Recorded                bool           `json:"recorded"`
	Profile                 string         `json:"profile"`
	Metadata                string         `json:"metadata"`
	RequiredTopics          []string       `json:"required_topics"`
	OptionalTopics          []string       `json:"optional_topics"`
	PresentTopics           []string       `json:"present_topics"`
	MessageCounts           map[string]int `json:"message_counts"`
	MissingRequiredTopics   []string       `json:"missing_required_topics"`
	ZeroCountRequiredTopics []string       `json:"zero_count_required_topics"`
	PresentOptionalTopics   []string       `json:"present_optional_topics"`
	MissingOptionalTopics   []string       `json:"missing_optional_topics"`
}

func ProfileTopics(path string) (Topics, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return Topics{}, err
	}
	var required []string
	var optional []string
	for _, rawLine := range strings.Split(string(data), "\n") {
		line := strings.TrimSpace(rawLine)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		parts := strings.Fields(line)
		if len(parts) < 2 {
			continue
		}
		switch parts[0] {
		case "required":
			required = append(required, parts[1])
		case "optional":
			optional = append(optional, parts[1])
		}
	}
	all := append(append([]string{}, required...), optional...)
	return Topics{Required: required, Optional: optional, All: all}, nil
}

func LoadMetadataCounts(path string) (map[string]int, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return MetadataCountsFromString(string(data)), nil
}

func MetadataCountsFromString(content string) map[string]int {
	counts := map[string]int{}
	topicPattern := regexp.MustCompile(`name: (/[^\n]+)`)
	countPattern := regexp.MustCompile(`message_count:\s*(\d+)`)
	matches := topicPattern.FindAllStringSubmatchIndex(content, -1)
	for index, match := range matches {
		topic := strings.TrimSpace(content[match[2]:match[3]])
		end := len(content)
		if index+1 < len(matches) {
			end = matches[index+1][0]
		}
		block := content[match[1]:end]
		countMatch := countPattern.FindStringSubmatch(block)
		if len(countMatch) != 2 {
			counts[topic] = 0
			continue
		}
		count, err := strconv.Atoi(countMatch[1])
		if err != nil {
			count = 0
		}
		counts[topic] = count
	}
	return counts
}

func ValidateProfile(profilePath string, metadataPath string, required []string, optional []string) (Validation, error) {
	counts, err := LoadMetadataCounts(metadataPath)
	if err != nil {
		return Validation{}, err
	}
	presentTopics := make([]string, 0, len(counts))
	for topic := range counts {
		presentTopics = append(presentTopics, topic)
	}
	sort.Strings(presentTopics)

	missingRequired := missingTopics(required, counts)
	zeroCountRequired := zeroCountTopics(required, counts)
	presentOptional := positiveCountTopics(optional, counts)
	missingOptional := missingTopics(optional, counts)

	return Validation{
		OK:                      len(missingRequired) == 0 && len(zeroCountRequired) == 0,
		Recorded:                true,
		Profile:                 profilePath,
		Metadata:                metadataPath,
		RequiredTopics:          append([]string{}, required...),
		OptionalTopics:          append([]string{}, optional...),
		PresentTopics:           presentTopics,
		MessageCounts:           counts,
		MissingRequiredTopics:   missingRequired,
		ZeroCountRequiredTopics: zeroCountRequired,
		PresentOptionalTopics:   presentOptional,
		MissingOptionalTopics:   missingOptional,
	}, nil
}

func missingTopics(topics []string, counts map[string]int) []string {
	var missing []string
	for _, topic := range topics {
		if _, exists := counts[topic]; !exists {
			missing = append(missing, topic)
		}
	}
	return missing
}

func zeroCountTopics(topics []string, counts map[string]int) []string {
	var zero []string
	for _, topic := range topics {
		if counts[topic] <= 0 {
			zero = append(zero, topic)
		}
	}
	return zero
}

func positiveCountTopics(topics []string, counts map[string]int) []string {
	var present []string
	for _, topic := range topics {
		if counts[topic] > 0 {
			present = append(present, topic)
		}
	}
	return present
}
